brave-cxf3/src/main/java/com/github/kristofa/brave/cxf3/TracerContext.java

```diff
@@ -0,0 +1,100 @@
+package com.github.kristofa.brave.cxf3;
+
+import com.github.kristofa.brave.Brave;
+import com.github.kristofa.brave.ServerSpan;
+import com.github.kristofa.brave.ServerSpanThreadBinder;
+import org.apache.cxf.message.Message;
+
+import java.util.concurrent.Callable;
+
+/**
+ * Used to handle {@link ServerSpan} retrieved from cxf message exchange.
+ */
+public class TracerContext {
+
+  private final ServerSpanThreadBinder serverSpanThreadBinder;
+  private final Message message;
+
+  TracerContext(Brave brave, Message message) {
+    this.serverSpanThreadBinder = brave.serverSpanThreadBinder();
+    this.message = message;
+  }
+
+  /**
+   * Wraps provided {@link Runnable and ensures same {@link ServerSpan} context as the thread from which the Runnable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Runnable}
+   * @return wrapped {@link Runnable}
+   */
+  public Runnable wrap(Runnable command) {
+    if (command == null) {
+      return null;
+    }
+
+    // get server span bind to current thread
+    final ServerSpan current = getCurrentServerSpan();
+    final ServerSpan serverSpan = getFromExchange();
+
+    return () -> {
+      try {
+        // When current span is not null, it means it was bind manually by developer
+        // or this wrapper is used in sync scenario.
+        // Bind server span from Exchange only when current span is null.
+        if (current == null) {
+          serverSpanThreadBinder.setCurrentSpan(serverSpan);
+        }
+        command.run();
+      } finally {
+        serverSpanThreadBinder.setCurrentSpan(current);
+      }
+    };
+  }
+
+  /**
+   * Wraps provided {@link Callable} and ensures same {@link ServerSpan} context as the thread from which the Callable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Callable}
+   * @return wrapped {@link Callable}
+   */
+  public <V> Callable<V> wrap(Callable<V> command) {
+    if (command == null) {
+      return null;
+    }
+
+    // get server span bind to current thread
+    final ServerSpan current = getCurrentServerSpan();
+    final ServerSpan serverSpan = getFromExchange();
+
+    return () -> {
+      try {
+        // When current span is not null, it means it was bind manually by developer
+        // or this wrapper is used in sync scenario.
+        // Bind server span from Exchange only when current span is null.
+        if (current == null) {
+          serverSpanThreadBinder.setCurrentSpan(serverSpan);
+        }
+        return command.call();
+      } finally {
+        serverSpanThreadBinder.setCurrentSpan(current);
+      }
+    };
+  }
+
+  private ServerSpan getCurrentServerSpan() {
+    // gets existing server span or reinitializes to ServerSpan#EMPTY using ThreadLocal#initialValue
```

Comment:
So if I understood correctly, `CurrentSpanContext` would look like this?

```java
public final class CurrentSpanContext {

  private final ServerSpanThreadBinder serverSpanThreadBinder;
  private final Message message;

  CurrentSpanContext(Brave brave, Message message) {
    this.serverSpanThreadBinder = brave.serverSpanThreadBinder();
    this.message = message;
  }

  public void attach() {
    // get server span bind to current thread
    final ServerSpan current = serverSpanThreadBinder.getCurrentServerSpan();

    // when no server span is attached used this from exchange
    if (current == null || ServerSpan.EMPTY == current) {
      final ServerSpan fromExchange = getFromExchange();
      serverSpanThreadBinder.setCurrentSpan(fromExchange);
    }
  }

  private <T> T getFromExchange() {
    return (T) message.getExchange().get(BraveCxfConstants.BRAVE_SERVER_SPAN);
  }
}
```

and used like this:

```java
public class FooEndpoint {
    private final Executor executor;

    public FooService(Executor executor) {
      this.executor = executor;
    }

    @POST
    @Path("/foo")
    public void foo(@Suspended AsyncResponse response, @Context CurrentSpanContext context) {
      executor.execute(() -> {
      	// attaches server span from exchange
        context.attach();
        
        response.resume("foo");
      });
    }
  }
```